### PR TITLE
3.x PaginatorHelper::defaultScope() and 'scope' / Consistency with PaginatorComponent

### PR DESCRIPTION
**Pull request by inoas**
Ref: https://github.com/cakephp/cakephp/issues/9211

### CODE REVIEW

**Comment by inoas** [comment]
I hope it is close to what you had in mind @markstory.

File: src/View/Helper/PaginatorHelper.php [hunk 1]
```diff
@@ -1106,4 +1153,20 @@ public function implementedEvents()
...
         return [];
     }
+
+    /**
+     * Helper method to deprecate the 'model' option throughout PaginationHelper.
+     *
+     * @param array $options Array of options
+     * @return array Array of options
+     */
+    private function _deprecateModelOption(array $options)
```
Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: Protected please.

**Comment by markstory** [comment]
Moving to 3.4 and retargeting to 3.next as this changes protected properties that could break userland code.

File: src/View/Helper/PaginatorHelper.php [hunk 2]
```diff
@@ -568,47 +582,60 @@ public function hasPage($model = null, $page = 1)
...
 
         return !empty($params) && $params[$page . 'Page'];
     }
 
     /**
-     * Gets or sets the default model of the paged sets
+     * Gets or sets the default scope of the paged sets
      *
-     * @param string|null $model Model name to set
+     * @deprecated use defaultScope() instead.
```
Comment (markstory) [inline src/View/Helper/PaginatorHelper.php]: Should have a version number here (3.4.0)

File: src/View/Helper/PaginatorHelper.php [hunk 3]
```diff
@@ -1106,4 +1153,21 @@ public function implementedEvents()
...
+     * Helper method to deprecate the 'model' option throughout PaginationHelper.
+     *
+     * @deprecated
+     * @param array $options Array of options
+     * @return array Array of options
+     */
+    protected function _deprecateModelOption(array $options)
+    {
+        if (isset($options['model'])) {
+            $options['scope'] = $options['model'];
```
Comment (markstory) [inline src/View/Helper/PaginatorHelper.php]: Perhaps scope should only be written if model is defined and scope is *not* defined?

File: tests/TestCase/View/Helper/PaginatorHelperTest.php [hunk 4]
```diff
@@ -2253,7 +2253,7 @@ public function testCounter()
         $expected = '1 - 3 of 13';
         $this->assertEquals($expected, $result);
 
-        $result = $this->Paginator->counter('Showing {{page}} of {{pages}} {{model}}');
+        $result = $this->Paginator->counter('Showing {{page}} of {{pages}} {{scope}}');
```
Comment (markstory) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php]: Does the old model variable still work?

File: src/View/Helper/PaginatorHelper.php [hunk 5]
```diff
@@ -158,34 +159,38 @@ public function options(array $options = [])
...
-            unset($options[$model]);
+            $this->request->params['paging'][$scope] = $options[$scope] + $this->request->params['paging'][$scope];
+            unset($options[$scope]);
         }
         $this->_config['options'] = array_filter($options + $this->_config['options']);
         if (empty($this->_config['options']['url'])) {
             $this->_config['options']['url'] = [];
         }
-        if (!empty($this->_config['options']['model'])) {
+        if (!empty($this->_config['options']['scope'])) {
```
Comment (ADmad) [inline src/View/Helper/PaginatorHelper.php]: To maintain BC using old key `model` should still be possible.

File: src/View/Helper/PaginatorHelper.php [hunk 6]
```diff
@@ -284,13 +289,13 @@ protected function _toggledLink($text, $enabled, $options, $templates)
...
         }
-        $paging = $this->params($options['model']);
+        $paging = $this->params($options['scope']);
 
         $url = array_merge(
             $options['url'],
             ['page' => $paging['page'] + $options['step']]
         );
-        $url = $this->generateUrl($url, $options['model']);
+        $url = $this->generateUrl($url, $options['scope']);
```
Comment (ADmad) [inline src/View/Helper/PaginatorHelper.php]: To maintain BC using old key `model` should still be possible.

**Review by ADmad — CHANGES_REQUESTED** [review]

File: tests/TestCase/View/Helper/PaginatorHelperTest.php (hunk 4 above)
Comment (inoas) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php]: Fixed that in a new commit.

File: src/View/Helper/PaginatorHelper.php (hunk 2 above)
Comment (inoas) [inline src/View/Helper/PaginatorHelper.php]: Fixed in new commit.

**Comment by inoas** [comment]
If you want me to I can copy the old tests in the testcases as well.

File: src/View/Helper/PaginatorHelper.php (hunk 3 above)
Comment (markstory) [inline src/View/Helper/PaginatorHelper.php]: How would userland code break on a parent call for a method that doesn't exist yet?

Comment (inoas) [inline src/View/Helper/PaginatorHelper.php]: Say someone has written code like:

```php
$this->Paginator->options(['model' => 'Client', 'scope' => 'ThinClient']);
```

... having overloaded Paginator with his own PaginatorHelper which extends upon the core PaginatorHelper.
Here the meaning of scope is/was a different.

I can change it to B, if you say so. I am just not sure it is entirely better.

Comment (markstory) [inline src/View/Helper/PaginatorHelper.php]: We don't make promises around backwards-compatibility if people override methods and add arguments/options.

File: src/View/Helper/PaginatorHelper.php [hunk 7]
```diff
@@ -1106,4 +1159,21 @@ public function implementedEvents()
     {
         return [];
     }
+
+    /**
+     * Helper method to deprecate the 'model' option throughout PaginationHelper.
+     *
+     * @deprecated 3.4.0
+     * @param array $options Array of options
+     * @return array Array of options
```
Comment (ravage84) [inline src/View/Helper/PaginatorHelper.php]: Either here or in the long description of this method you could/should document the intent a little bit more.
A one liner should suffice. What is been done to the model option to deprecate it?

File: tests/TestCase/View/Helper/PaginatorHelperTest.php [hunk 8]
```diff
@@ -2253,6 +2253,10 @@ public function testCounter()
         $expected = '1 - 3 of 13';
         $this->assertEquals($expected, $result);
 
+        $result = $this->Paginator->counter('Showing {{page}} of {{pages}} {{scope}}');
+        $this->assertEquals('Showing 1 of 5 clients', $result);
+
+        // Deprecated 3.4.0:
```
Comment (ravage84) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php]: "Deprecated as of 3.4.0"

**Review by ravage84 — CHANGES_REQUESTED** [review]

**Comment by dereuromark** [comment]
Is this still interesting to be finished up? @inoas

**Comment by inoas** [comment]
I really do want to finish the few PRs I have opened yet. If you want to close them as part of clean up - feel free, I will get back to them as soon as project pressure is less.

**Comment by dereuromark** [comment]
This would now be able to go into 3.6
Are you still interested in finishing it up? Otherwise we should close.

**Comment by inoas** [comment]
Do I see correctly that aside the merge conflict and some CS checks the unit tests already run?

I think I can pick this up the coming weeks (after doing some more work on TwigView) and finish this in 3.6 so that the deprecated name can be removed in 4.0...

**Pull request closed without merging**